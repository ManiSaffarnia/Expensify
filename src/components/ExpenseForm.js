import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            errorState: ''
        };
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
        else {
            console.log("doesn't match");
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };


    onSubmit = (e) => {
        if (!this.state.description || !this.state.amount) {
            //error ke khali nabashe field ha
            this.setState(() => ({ errorState: "Please provide description and amount" }))
        }
        else {
            this.setState(() => ({ errorState: '' }));
            //expense jadid vared shode ro barmigardoonim be bala
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
        e.preventDefault();
    };



    render() {
        return (
            <div className="content-container">
                <form className="form" onSubmit={this.onSubmit}>
                    {this.state.errorState && <p className="form__error" id="errorState">{this.state.errorState}</p>}
                    <input className="text-input" type="text" name="description" placeholder="Description" autoFocus value={this.state.description} onChange={this.onDescriptionChange} />
                    <input className="text-input" type="text" name="amount" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange} />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea className="text-area" name="note" id="" cols="30" rows="10" placeholder="Add a note for yout expense (optional)" value={this.state.note} onChange={this.onNoteChange} />
                    <div>
                        <button className="button">Save Expense</button>
                    </div>
                </form>
            </div>
        )
    }
}