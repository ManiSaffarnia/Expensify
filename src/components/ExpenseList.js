import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => {
    return (
        <div>
            <h1>
                Expense List
            </h1>

            {
                props.expenses.length > 0 ? (
                    props.expenses.map((expense) => (
                        <ExpenseListItem {...expense} key={expense.id} />
                    ))
                    ) : (
                    <p>There is no Expense</p>
                    )
            }
        
        </div>
    );
};

const mapStateToProps = (state) => {
    //console.log(state);
    return ({
        expenses: selectExpenses(state.expenses, state.filters)
    });
};

export default connect(mapStateToProps)(ExpenseList);