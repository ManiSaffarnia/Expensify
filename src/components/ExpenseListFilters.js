import React from 'react';
import { connect } from 'react-redux';
import { sortByAmount, sortByDate, setStartDate, setEndDate, setTextFilter } from "../actions/filters";
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSelectChange = (e) => {
        (e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount());
    };

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input className="text-input" placeholder="Search expenses" type="text" value={this.props.filters.text} onChange={this.onTextChange} />
                    </div>
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.onSelectChange}
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId={"123"}
                            endDate={this.props.filters.endDate}
                            endDateId={"321"}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            //options
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </div>
                </div>



            </div>
        );
    }
}


const mapStateToProps = (state) => {
    return ({
        filters: state.filters
    });
};

const mapDispatchToProps = (dispatch, props) => ({
    setTextFilter: (data) => (dispatch(setTextFilter(data))),
    setStartDate: (startDate) => (dispatch(setStartDate(startDate))),
    setEndDate: (endDate) => (dispatch(setEndDate(endDate))),
    sortByDate: () => (dispatch(sortByDate())),
    sortByAmount: () => (dispatch(sortByAmount()))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);