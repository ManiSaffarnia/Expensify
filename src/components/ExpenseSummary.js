import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import expenseTotal from '../selectors/expense-total';
import selectExpenses from '../selectors/expenses';
import { Link } from 'react-router-dom';

const ExpenseSummary = ({ expensesCount, expensesTotal }) => {
    const totalAmount = numeral(expensesTotal / 100).format('$0,0.00');

    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span> {expensesCount} </span> {expensesCount > 1 ? 'expenses' : 'expense'} totalling <span> {totalAmount} </span></h1>
                <div className="page-header__actions">
                    <Link to="/create" className="button"> Add Expense </Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    const filteredExpense = selectExpenses(state.expenses, state.filters);
    return ({
        expensesCount: filteredExpense.length,
        expensesTotal: expenseTotal(filteredExpense)
    });
};

export default connect(mapStateToProps)(ExpenseSummary);