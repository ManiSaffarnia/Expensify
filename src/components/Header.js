import React from 'react';
import {NavLink} from 'react-router-dom';
import {startLogOut} from '../actions/auth';
import {connect} from "react-redux";

const Header = (props) => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/dashboard" activeClassName="is-active" exact={true}>Dashboard</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help</NavLink>
        <button onClick={props.dispatch(startLogOut)}>Sign Out</button>
    </header>
);

export default connect()(Header);