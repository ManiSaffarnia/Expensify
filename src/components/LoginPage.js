import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

const Login = (props) => {
    return (
        <div>
            <p>Login Page</p>
            <button onClick={props.dispatch(startLogin)}>Login</button>
        </div>
    )
};

export default connect()(Login);