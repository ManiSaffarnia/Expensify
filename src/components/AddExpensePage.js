import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from "../actions/expenses";

const AddExpensePage = (props) => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={(expense) => {
            //console.log(expense);
            props.dispatch(startAddExpense({
                description: expense.description,
                note: expense.note,
                amount: expense.amount,
                createdAt: expense.createdAt
            }));
            props.history.push('/');
        }} />
    </div>
);

export default connect()(AddExpensePage);