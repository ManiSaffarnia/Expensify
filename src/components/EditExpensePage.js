import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    }

    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push("/");
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit} />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        );
    }
}


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            if (expense.id === props.match.params.id) {
                return true;
            }
        })
    }
};

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (expenseId, newExpense) => dispatch(startEditExpense(expenseId, newExpense)),
    startRemoveExpense: (expense) => dispatch(startRemoveExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);