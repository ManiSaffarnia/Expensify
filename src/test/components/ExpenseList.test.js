import React from 'react';
import { shallow } from 'enzyme'
import {ExpenseList} from '../../components/ExpenseList';
import expensesTestData from '../fixture/expenses';

test('should render ExpenseList component with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expensesTestData}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList component with no Expense', () => {
    const wrapper = shallow(<ExpenseList expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
});