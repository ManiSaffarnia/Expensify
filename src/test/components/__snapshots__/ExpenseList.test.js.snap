// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList component with expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id={1}
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={299500}
    createdAt={-345600000}
    description="Rent"
    id={2}
    key="2"
    note="August rent"
  />
  <ExpenseListItem
    amount={5000}
    createdAt={345600000}
    description="food"
    id={3}
    key="3"
    note="yesterday dinnet"
  />
</div>
`;

exports[`should render ExpenseList component with no Expense 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    There is no Expense
  </p>
</div>
`;
